bigchaindb/elections/election.py

```diff
@@ -242,24 +242,32 @@ def show_election(self, bigchain):
         return response
 
     @classmethod
-    def approved_update(cls, bigchain, new_height, txns):
-        votes = {}
+    def approved_elections(cls, bigchain, new_height, txns):
+        elections = {}
         for txn in txns:
             if not isinstance(txn, Vote):
                 continue
 
             election_id = txn.asset['id']
-            election_votes = votes.get(election_id, [])
+            e = bigchain.get_transaction(election_id)
+            election_operation = e.OPERATION
+            election_operation_votes = elections.get(election_operation, {})
+            # Once we conclude an election of a given type, we stop looking at that election class
+            if type(election_operation_votes) != dict:
+                continue
+            election_votes = election_operation_votes.get(election_id, [])
             election_votes.append(txn)
-            votes[election_id] = election_votes
+            election_operation_votes[election_id] = election_votes
+            elections[election_operation] = election_operation_votes
 
             election = cls.has_concluded(bigchain, election_id, election_votes, new_height)
-            # Once an election concludes any other conclusion for the same
-            # or any other election is invalidated
             if election:
+                # Once we conclude an election, we store the result in the db
                 cls.store_election_results(bigchain, election, new_height)
-                return cls.on_approval(bigchain, election, new_height)
-        return None
+                # And keep the transaction filed under the election_type
+                elections[election_operation] = election.on_approval(bigchain, election, new_height)
```

Comment:
We need to be able to foresee if a particular election updates the validator set and only persist (conclude) one of those if there are more than one. Elections that do not update the validator set are concluded too.

Reply:
This is roughly how the conclusion can look like:
```
votes_by_election = defaultdict(list)
for txn in txns:
    if not isinstance(txn, Vote):
        continue
    election_id = txn.asset['id']
    votes_by_election[election_id].append(tx)

updated_validator_set = False  # only one validator set update is allowed per block
for election_id, votes in votes_by_election:
    election = Election.from_dict(bigchain.get_transaction(election_id))

    if not election.has_concluded(bigchain, election_id, votes, new_height):
        continue

    if election.get_validator_set() != bigchain.get_validators():
        if updated_validator_set:
            continue
        updated_validator_set = True

    election.on_approval(bigchain, new_height)
```
What do you think?

Reply:
I wanted to talk to you about this tomorrow.

The logic as I have it now allows for only one concluded election of a given type per block. This is desirable for `ValidatorElection`s, and doesn't really matter for `MigrationElection`s, but what do we want the generalized behavior to be?

Also, I think we can't easily avoid type-checking the `election`s. `Election.get_validators()` just gets the validator set from `bigchain`. The only place a `ValidatorElection` stores the new validator set is in its asset data. So it's probably easiest just to check the `OPERATION` or the class `.__name__`.

Reply:
We have no reason to have a one per block restriction when it's not about the validator set update.

A way to set elections that update validator set apart is to introduce a generic method that reports the new validator set. It can return an empty list by default whereas `ValidatorElection` will infer it from the asset.